File: java/akka-persistence/counter-sample_protobuf/src/main/java/sample/SampleActor.java
package sample;

import akka.japi.pf.ReceiveBuilder;
import akka.persistence.AbstractPersistentActor;
import akka.persistence.SnapshotOffer;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

import sample.Count.AddedCounter;

public class SampleActor extends AbstractPersistentActor {
    private final String id;
    private int state = 0;

    public SampleActor(String id) {
        this.id = id;
    }

    @Override
    public String persistenceId() {
        return id;
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receiveRecover() {
        return ReceiveBuilder
                .match(AddedCounter.class, this::updateState)
                .match(SnapshotOffer.class, ss -> state = (Integer)ss.snapshot())
                .build();
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receiveCommand() {
        return ReceiveBuilder
                .match(CounterAdd.class, cmd ->
                        persist(createEvent(cmd), event -> {
                            updateState(event);
                            context().system().eventStream().publish(event);
                        })
                )
                .matchEquals("snapshot", cmd -> saveSnapshot(state))
                .matchEquals("dump", cmd -> System.out.println("counter: " + state))
                .matchEquals("terminate", cmd -> context().system().terminate())
                .build();
    }

    private void updateState(AddedCounter event) {
        state += event.getCount();
    }

    private AddedCounter createEvent(CounterAdd cmd) {
        return AddedCounter.newBuilder()
                .setCount(cmd.count())
                .build();
    }
}

File: scala/akka-persistence/query-sample/src/main/scala/sample/SampleActor.scala
package sample

import akka.persistence._

case class CounterAdd(count: Int)
case class AddedCounter(count: Int)

class SampleActor extends PersistentActor {
  var state = 0

  def updateState(event: AddedCounter): Unit = {
    state += event.count
  }

  override def persistenceId: String = "counter"

  override def receiveRecover: Receive = {
    case ev: AddedCounter => {
      println(s"*** receiveRecover - event: ${ev}")

      updateState(ev)
    }
    case SnapshotOffer(_, snapshot: Int) => {
      println(s"*** receiveRecover - snapshot: ${snapshot}")

      state = snapshot
    }
  }

  override def receiveCommand: Receive = {
    case cmd: CounterAdd => persist(AddedCounter(cmd.count)) { event =>
      println("*** persist event")

      updateState(event)
      context.system.eventStream.publish(event)
    }
    case "snapshot" => saveSnapshot(state)
    case "dump" => println(s"counter: ${state}")
    case "end" => context.stop(self)
  }
}
